### PR TITLE
feat: Add option to change the log level of the logs emitted by triggers

### LINKED ISSUE #8329: Add option to change the log level of the logs emitted by triggers
### New Feature / Enhancement Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Current Limitation
Parse Server can really get very wordy when actively using triggers (beforeSave, afterSave) : this can quickly lead to oversized log files and drowned informations.

By default, theses messages are written to the log file using the 'info' channel. The only way to shut them off is to raise the log level of Parse Server to the 'warn' level.

Customization is a key characteristic of Parse Server.  Developers should be able to modify the channel so they can enforce their log policy.

### Feature / Enhancement Description

Add three log level options into the Parse Server config (logLevelTriggerAfterHook, logLevelTriggerSuccessBeforeHook, logLevelTriggerErrorBeforeHook) in order to fine-tune the triggers log channel according Winston's one ('info', 'debug','warn').

### Example Use Case

1. Add "logLevelUses: { triggerAfterHook: 'debug' }" to Parse Server configuration
2. Save a Parse Object
3. The log is not written on the 'info' channel anymore, but use the 'debug'.

### Alternatives / Workarounds

The only way to shut them off is to raise the log level of Parse Server to the 'warn' level.

### PR DESCRIPTION
**Pull request by alljinx**
### New Pull Request Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description

Related issue: https://github.com/parse-community/parse-server/issues/8329
Closes: https://github.com/parse-community/parse-server/issues/8329

Parse Server can really get very wordy when actively using triggers (beforeSave, afterSave) : this can quickly lead to oversized log files and drowned informations.

By default, theses messages are written to the log file using the 'info' channel. The only way to shut them off is to raise the log level of Parse Server to the 'warn' level.

Customization is a key characteristic of Parse Server.  Developers should be able to modify the channel so they can enforce their log policy.

### Approach

Add three log level options into the Parse Server config (logLevelTriggerAfterHook, logLevelTriggerSuccessBeforeHook, logLevelTriggerErrorBeforeHook) in order to fine-tune the triggers log channel according Winston's one ('info', 'debug','warn').

### TODOs before merging

- [x] Add tests
- [x] Add changes to documentation (guides, repository pages, in-code descriptions)

### CODE REVIEW

File: src/Options/Definitions.js
Comment (mtrezza) [inline src/Options/Definitions.js]: This is an auto-generated file, was this generated via "npm run definitions"?

File: src/triggers.js
Comment (mtrezza) [inline src/triggers.js]: I'm not sure about these references, could there be cases where `config.logLevelUses` is `undefined`?

File: src/Options/index.js
Comment (mtrezza) [inline src/Options/index.js]:
```suggestion
export interface LogLevels {
```

File: src/Options/index.js
Comment (mtrezza) [inline src/Options/index.js]:
```suggestion
  /* (Optional) Overrides the log levels used internally by Parse Server to log events. */
  logLevels: ?LogLevels;
```

I'd simplify the name and remove the "uses". The `:ENV:`name  should be generated automatically when you run `npm run definitions`.

File: src/Options/index.js
Comment (mtrezza) [inline src/Options/index.js]: `beforeSave`, `afterSave`, etc. are called "Parse Cloud Triggers". "Hook" could be mistaken for "Cloud Code Webhook", which is a different feature.

For consistency maybe we name the keys
- `triggerAfter`
- `triggerBeforeSuccess`
- `triggerBeforeError`

In the future someone could add names like
- `triggerBeforeLoginSuccess`
- `triggerBeforeGetFileError`

File: src/Options/index.js
Comment (mtrezza) [inline src/Options/index.js]:
```suggestion
  /* Log level used by the Cloud Code Triggers `afterSave`, `afterDelete`, `afterSaveFile`, `afterDeleteFile`, `afterFind`, `afterLogout`. Default is `info`.
```

I think we should name each trigger individually to avoid confusion. There is the `afterLiveQueryEvent` but I'm not sure whether it is within the scope of this trigger log function for example.

File: src/Options/index.js
Comment (mtrezza) [inline src/Options/index.js]:
```suggestion
  /* Log level used by the Cloud Code Triggers `beforeSave`, `beforeSaveFile`, `beforeDeleteFile`, `beforeFind`, `beforeLogin` on success. Default is `info`.
```

Not sure whether LQ triggers `beforeConnect`, `beforeSubscribe` are also covered by this?

File: src/Options/index.js
Comment (mtrezza) [inline src/Options/index.js]:
```suggestion
  /* Log level used by the Cloud Code Triggers `beforeSave`, `beforeSaveFile`, `beforeDeleteFile`, `beforeFind`, `beforeLogin` on success. Default is `error `.
```

Again, not sure whether this also covers LQ triggers.

File: spec/CloudCodeLogger.spec.js
Comment (mtrezza) [inline spec/CloudCodeLogger.spec.js]: could you refactor this test, it contains two identical blocks; instead you could just write a list of levels and test them in a loop?

**Review by mtrezza — CHANGES_REQUESTED** [review]

File: src/Options/Definitions.js
Comment (alljinx) [reply src/Options/Definitions.js]: Oh... My bad, I didn't know this is an auto-generated file. Last commit fix this.

File: src/triggers.js
Comment (alljinx) [reply src/triggers.js]: I got confused about how to pass a default value. Last commit fix this.

File: src/Options/index.js
Comment (alljinx) [reply src/Options/index.js]: Last commit fix this.

File: src/Options/index.js
Comment (alljinx) [reply src/Options/index.js]: Got it, last commit fix this.

File: spec/CloudCodeLogger.spec.js
Comment (alljinx) [reply spec/CloudCodeLogger.spec.js]: Commited a fix, should be better now.

File: src/Options/Definitions.js
Comment (mtrezza) [reply src/Options/Definitions.js]: Yes, you can find more info here: https://github.com/parse-community/parse-server/blob/alpha/CONTRIBUTING.md#parse-server-configuration

File: resources/buildConfigDefinitions.js
Comment (mtrezza) [inline resources/buildConfigDefinitions.js:110]: Could you run `npm run prettier && npm run lint-fix`? It seems there were some style changes unrelated to this PR.

**Review by mtrezza — CHANGES_REQUESTED** [review]

File: resources/buildConfigDefinitions.js
Comment (alljinx) [reply resources/buildConfigDefinitions.js:110]: Done !

**Review by mtrezza — COMMENTED** [review]
Did you make any manual changes at all to `resources/buildConfigDefinitions.js` except for adding `'LogLevels',`, or are these all automatic style changes?

File: src/Options/index.js
Comment (mtrezza) [inline src/Options/index.js:3]: Could you revert this reordering of import statements? We want to keep only PR changes that are necessary for the PR itself.

File: src/triggers.js
Comment (mtrezza) [inline src/triggers.js]: Would it be better to pass the log level `none` also to the logger and handle it there? This way we don't need the `if` statement everywhere we log something.

File: src/Options/Definitions.js
Comment (mtrezza) [inline src/Options/Definitions.js]: `undefinedTRIGGER_AFTER` something went wrong there, and in the ones below; I think the answer is in the contribution docs

File: spec/CloudCodeLogger.spec.js
Comment (mtrezza) [inline spec/CloudCodeLogger.spec.js]:
```suggestion
      spy = spyOn(Config.get('test').loggerController.adapter, 'log').and.callThrough();
      const obj = new Parse.Object('TestClass');
      await obj.save();
      return spy;
```

File: spec/CloudCodeLogger.spec.js
Comment (mtrezza) [inline spec/CloudCodeLogger.spec.js]:
```suggestion
  it('should log cloud function triggers using the custom log level', async () => {
```

File: spec/CloudCodeLogger.spec.js
Comment (mtrezza) [inline spec/CloudCodeLogger.spec.js]:
```suggestion
```

File: src/Config.js
Comment (mtrezza) [inline src/Config.js]: To simplify, can you just import controller and reference it directly below where you need its logLevels property?
 
```suggestion
import { logLevels as validLogLevels } from './Controllers/LoggerController';
```

**Review by mtrezza — CHANGES_REQUESTED** [review]

File: src/Config.js
Comment (alljinx) [reply src/Config.js]: Sorry, but I'm not sure about how you want me to do that : ```logLevels``` is a not property of LoggerController but a constant...

File: src/triggers.js
Comment (alljinx) [reply src/triggers.js]: I think you're right, it would be better to multualize but I'm not sur it's the right way : It seems weird to me to add a ```none()``` function that do nothing.
Maybe I was just wrong in my analysis and the 'none' is useless : if you don't want to save a log, you just have to switch the level of the trigger to 'silly' or 'debug'. What do you think ?

File: src/Options/Definitions.js
Comment (alljinx) [reply src/Options/Definitions.js]: Fixed !

File: src/triggers.js
Comment (alljinx) [reply src/triggers.js]: Just commit a fix in this way, I think it's more consistent.

Comment (mtrezza) [reply src/triggers.js]: Some log level definitions have a concept of `none` or `off`. From the logger's perspective it seems odd, because the logger should be called when there's something to log. From the developer's perspective it makes more sense, because they want to control whether something is logged at all or if yes at which log level.

We can solve this in 2 ways:
- a) Never call the logger directly anywhere in code; instead add a proxy method that which includes the `if "none", don't log" condition.
- b) Just add this condition to the logger itself, to allow passing the developer's choice directly to the logger.

Option (a) seems to imply a lot of code changes, and we have no easy way to enforce not calling the logger directly when reviewing PRs. So I'd go with option (b).

File: src/Config.js
Comment (mtrezza) [inline src/Config.js]:
```suggestion
          throw `'${key}' must be one of ${JSON.stringify(validLogLevels)}`;
```

If this is an array, maybe we need to stringify for proper log output?

File: src/Config.js
Comment (mtrezza) [reply src/Config.js:9]: Got it, I'd just give it a more descriptive name then, like `allowedLogLevels`

Comment (mtrezza) [reply src/Config.js]: Got it, then I would just give it a more descriptive name like `validLogLevels`

File: src/Options/Definitions.js
Comment (mtrezza) [inline src/Options/Definitions.js:909]: Was `PARSE_SERVER_LOG_LEVELS_TRIGGER_AFTER` created running `npm run definitions`, or did you manually add this?

File: src/Controllers/LoggerController.js
Comment (mtrezza) [inline src/Controllers/LoggerController.js:19]: So we don't offer a `none` log level in this PR to disable logging for a specific log entry type, correct?

**Review by mtrezza — COMMENTED** [review]
Could you look at the failing tests? They seem to be for the Postgres adapter.

File: src/Options/Definitions.js
Comment (alljinx) [reply src/Options/Definitions.js:909]: I used ```npm run definitions```

File: src/Controllers/LoggerController.js
Comment (alljinx) [reply src/Controllers/LoggerController.js:19]: Correct, I think there is enough levels to handle logs properly.
For example, in my company, we would use the 'debug' level for triggers because we feel this is not relevant for production. And this is really the point of this PR : prevent the log file to grow dramatically with useless informations that could drown out those that actually are.
Moreover, if this is not enough, it remains the 2 lower levels 'verbose' and 'silly', wich are, I guess, perfectly for that.
So, adding one more option seems to be a bit overkill and I would suggest to stay simple.

**Comment by alljinx** [comment]
> Could you look at the failing tests? They seem to be for the Postgres adapter.

Off course I will...

File: src/Controllers/LoggerController.js
Comment (mtrezza) [reply src/Controllers/LoggerController.js:19]: Sounds reasonable, good choice I think

**Comment by alljinx** [comment]
> Could you look at the failing tests? They seem to be for the Postgres adapter.

Fixed !

**Comment by alljinx** [comment]
Miss click, sorry...

**Comment by alljinx** [comment]
Hi, what's the next step plz ? Do you want me to do something more or are we just waiting for a another maintainer ?

**Review by mtrezza — APPROVED** [review]
Looks good! Just to confirm, if you run `npm run definitions` there are no code changes because you did not manually edit the `src/Options/docs.js` or `src/Options/Definitions.js` files, correct?

**Comment by alljinx** [comment]
> Looks good! Just to confirm, if you run `npm run definitions` there are no code changes because you did not manually edit the `src/Options/docs.js` or `src/Options/Definitions.js` files, correct?

Correct, I used `npm run definitions` :)

**Comment by mtrezza** [comment]
Alright, let's wait for the CI to pass and I think then we can merge it

**Comment by alljinx** [comment]
Everything seems ok to me since the last failing test is not related to my PR 👍

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0-alpha.13](https://github.com/parse-community/parse-server/releases/tag/6.0.0-alpha.13)

**Comment by SebC99** [comment]
I haven't spent a lot of time looking at the code changes, but one thing I noticed before and that may still be here is that Parse stringify the input / output of the triggers before evaluating the logLevel. So even when logs are not going to be logged, the stringify operation occurs, which is one of the most heavy operation of Parse.
It would be awesome to improve this.

**Comment by mtrezza** [comment]
Interesting observation, could you open a new issue for this?

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/6.0.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0](https://github.com/parse-community/parse-server/releases/tag/6.0.0)

**Pull request merged**
